examples/frost.c

```diff
@@ -0,0 +1,236 @@
+#include <secp256k1_frost.h>
+
+ /* Number of public keys involved in creating the aggregate signature */
+#define N_SIGNERS 5
+
+ /* Threshold required in creating the aggregate signature */
+#define THRESHOLD 3
+
+struct signer_secrets {
+    secp256k1_keypair keypair;
+    secp256k1_frost_share agg_share;
+    secp256k1_frost_secnonce secnonce;
+};
+
+struct signer {
+    secp256k1_xonly_pubkey pubkey;
+    secp256k1_frost_pubnonce pubnonce;
+    secp256k1_frost_partial_sig partial_sig;
+    secp256k1_pubkey pubcoeff[THRESHOLD];
+    unsigned char vss_hash[32];
+};
+
+ /* Create a key pair and store it in seckey and pubkey */
+int create_keypair(const secp256k1_context* ctx, struct signer_secrets *signer_secrets, struct signer *signer) {
+    unsigned char seckey[32];
+    FILE *frand = fopen("/dev/urandom", "r");
+    if (frand == NULL) {
+        return 0;
+    }
+    do {
+        if(!fread(seckey, sizeof(seckey), 1, frand)) {
+             fclose(frand);
+             return 0;
+         }
+    /* The probability that this not a valid secret key is approximately 2^-128 */
+    } while (!secp256k1_ec_seckey_verify(ctx, seckey));
+    fclose(frand);
+    if (!secp256k1_keypair_create(ctx, &signer_secrets->keypair, seckey)) {
+        return 0;
+    }
+    if (!secp256k1_keypair_xonly_pub(ctx, &signer->pubkey, NULL, &signer_secrets->keypair)) {
+        return 0;
+    }
+    return 1;
+}
+
+ /* Create shares and coefficient commitments */
+int create_shares(const secp256k1_context* ctx, struct signer_secrets *signer_secrets, struct signer *signer, secp256k1_xonly_pubkey *agg_pk) {
+    int i;
+    secp256k1_frost_share shares[N_SIGNERS][N_SIGNERS];
+    const secp256k1_pubkey *pubcoeffs[N_SIGNERS];
+
+    for (i = 0; i < N_SIGNERS; i++) {
+        /* Generate a polynomial share for each participant */
+        if (!secp256k1_frost_share_gen(ctx, signer[i].pubcoeff, shares[i], THRESHOLD, N_SIGNERS, &signer_secrets[i].keypair)) {
+            return 0;
+        }
+        pubcoeffs[i] = signer[i].pubcoeff;
+    }
+
+    /* KeyGen communication round 1: exchange shares, nonce commitments, and
+     * coefficient commitments */
+    for (i = 0; i < N_SIGNERS; i++) {
+        int j;
+        const secp256k1_frost_share *assigned_shares[N_SIGNERS];
+
+        /* Each participant receives a share from each participant (including
+         * themselves) corresponding to their index. */
+        for (j = 0; j < N_SIGNERS; j++) {
+            assigned_shares[j] = &shares[j][i];
+        }
+        /* Each participant aggregates the shares they received. */
+        if (!secp256k1_frost_share_agg(ctx, &signer_secrets[i].agg_share, agg_pk, signer[i].vss_hash, assigned_shares, pubcoeffs, N_SIGNERS, THRESHOLD, i+1)) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+/* Sign the VSS commitments */
+int sign_vss(const secp256k1_context* ctx, struct signer_secrets *signer_secrets, struct signer *signer, unsigned char sigs[N_SIGNERS][64]) {
```

Comment:
What do we achieve by signing the `vss_hash`? Aren't we assuming secure channels between the signers already (to distribute the shares)?

Reply:
The `vss_hash` signatures address the requirement of a broadcast channel for VSS. To fulfill this requirement, each participant must verify that they received the same VSS commitment as each other participant, and we can do this by having each participant verify that each other participant signed the same `vss_hash`.

Reply:
That's a great point, the participants can use their existing secure channels to distribute the `vss_hash`.

Reply:
It does seem a little more efficient though, because we already need to distribute signatures for the PoKs, so by signing the `vss_hash`, which also serves as the PoK`, we avoid the requirement of each participant needing to distribute a `vss_hash` to each other participant.

Reply:
Oh, I had missed that the signatures serve as a PoK. Good point that only sending the signatures is therefore more efficient.